### PR TITLE
Remove dependency on spectral-cube

### LINKED ISSUE #922: Remove SpectralCube (Was: Build collapse method for Spectrum1D objects)
**Reporter: Jesse Averbukh**

[🐱](https://jira.stsci.edu/browse/JDAT-1835) [🐱](https://jira.stsci.edu/browse/JDAT-2112)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to remove dependency on `spectral-cube` but to do that we also need to refactor the `Collapse` plugin.

Fixes #922

Blocked by:

- [x] Decision on `Collapse` refactoring, e.g., do we even need to collapse to get spatial-spectral data?
* astropy/specutils#909
- [x] New `specutils` release beyond v1.5.0 for `Spectrum1D.collapse()`.
- [x] Fix CI

Blocks:

* #1040 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 1]
```diff
@@ -130,35 +132,44 @@ def vue_collapse(self, *args, **kwargs):
...
             spec_min = float(self.spectral_min) * u.Unit(self.spectral_unit)
             spec_max = float(self.spectral_max) * u.Unit(self.spectral_unit)
-            spec = spec.spectral_slab(spec_min, spec_max)
-
-        collapsed_spec = getattr(spec, self.selected_func.lower())(
-            axis=self.selected_axis)
-
-        data = Data(coords=collapsed_spec.wcs)
-        data['flux'] = collapsed_spec.filled_data[...]
+            spec = spectral_slab(spec, spec_min, spec_max)
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:135]: This requires resolution of astropy/specutils#909 upstream and then a new `specutils` release including that fix.

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 2]
```diff
@@ -130,35 +132,44 @@ def vue_collapse(self, *args, **kwargs):
...
+
+        # FIXME: Does not work.
+        # collapsed_spec = spec.collapse(self.selected_func.lower(), axis=self.selected_axis)
+        # data = Data(coords=collapsed_spec.wcs)
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:139]: I am not sure how to use `specutils.Spectrum1D.collapse` to get spatial-spectral object correctly here. Particularly, the axis order might not be the same as `spectral-cube`; not sure, as the warnings about flipping axis from `specutils` is very confusing.

But more importantly, do we really need so many options for axis if people only ever wants to collapse to get spatial-spatial?

Comment (rosteen) [reply jdaviz/configs/default/plugins/collapse/collapse.py:139]: In offline discussion we decided that giving "spectral" and "spatial" collapse options rather than numerical axes is the way to go here. None of us could think of a use case for collapsing only one spatial axis. FYI the specutils collapse method should take those two strings as valid `axis` inputs, so it is hopefully easy to implement.

Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/collapse.py:139]: As discussed in 2021-12-16 tag-up, "spectral" collapse already happens automatically on load, so having it also in the plugin would be redundant.

**Comment by rosteen** [comment]
@pllim I'm hoping that https://github.com/astropy/specutils/pull/911 resolves the `spectral_slab` blocker for this.

File: jdaviz/configs/default/plugins/collapse/collapse.py (hunk 2 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/collapse.py:139]: I forgot what's the conclusion here. Do we still support "spectral" collapse or just get rid of it?

Comment (rosteen) [reply jdaviz/configs/default/plugins/collapse/collapse.py:139]: I would say leave it in :shrug:

Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/collapse.py]: I cannot make sense of the previous implementation using spectral-cube, so I removed "axis" option altogether. #yolo

File: docs/known_bugs.rst [hunk 3]
```diff
@@ -111,11 +111,6 @@ try a different OS/browser combo.
 Specviz
 -------
 
-Collapse Plugin spectral bounds do not match selected region
```
Comment (pllim) [inline docs/known_bugs.rst:114]: This entry is misleading because Specviz does not load Collapse and there is already a similar entry above for Cubeviz.

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 4]
```diff
@@ -16,28 +14,15 @@
 __all__ = ['Collapse']
 
 
-spaxel = u.def_unit('spaxel', 1 * u.Unit(""))
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:19]: This is not used. Not sure why it is here. It is also defined in at least two other places in this package but removing those is out of scope here.

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 5]
```diff
@@ -108,14 +88,14 @@ def vue_list_subsets(self, event):
         temp_dict = {}
         # Attempt to filter out spatial subsets
         for key, region in temp_subsets.items():
-            if type(region) == RectanglePixelRegion:
+            if type(region) == SpectralRegion:
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:91]: I think this was overlooked in #1046 ?

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 6]
```diff
@@ -148,28 +125,26 @@ def vue_collapse(self, *args, **kwargs):
...
             f"Data set '{self._selected_data.label}' collapsed successfully.",
             color="success",
             sender=self)
         self.hub.broadcast(snackbar_message)
 
-        if self.selected_axis == 0 and self.selected_viewer != 'None':
+        # Spatial-spatial image only.
+        if self.selected_viewer != 'None':
             # replace the contents in the selected viewer with the results from this plugin
             self.app.add_data_to_viewer(self.viewer_to_id.get(self.selected_viewer),
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:149]: @kecnry , I noticed something weird where while the collapsed image shows in the selected viewer, I am unable to use Layers menu to change the display. I have to deselect it from Data and re-select for Layers to work again. Not sure if this is a known issue or not.

Also, the spectral Subset somehow colors the whole collapsed image in the tint of the Subset color. I can disable it by clicking the eye icon on the Layers dropdown. I think this is a separate known issue already.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py [hunk 7]
```diff
@@ -18,39 +18,12 @@ def test_linking_after_collapse(spectral_cube_wcs):
...
     coll.selected_data_item = 'test'
 
-    coll.selected_axis = 0
     coll.add_replace_results = False
     coll.vue_collapse()
 
     assert len(dc) == 2
     assert dc[1].label == 'Collapsed 1 test'
-    assert len(dc.external_links) == 3
+    assert len(dc.external_links) == 2
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/tests/test_collapse.py:26]: The results changed here but I don't know what it means. Plugin seems to work though. 🤷

File: jdaviz/tests/test_subsets.py [hunk 8]
```diff
@@ -70,8 +70,6 @@ def test_region_from_subset_profile(jdaviz_app, spectral_cube_wcs):
     subsets = jdaviz_app.get_subsets_from_viewer('spectrum-viewer', subset_type='spectral')
     reg = subsets.get('Subset 1')
 
-    print(reg)
```
Comment (pllim) [inline jdaviz/tests/test_subsets.py:73]: Leftover debugging from #1046 .

**Comment by javerbukh** [comment]
While testing this, I noticed that collapsed data is in the wrong orientation. Is this a working as expected or a bug?
Screen Shot 2022-02-08 at 8 26 31 AM

File: jdaviz/configs/default/plugins/collapse/collapse.vue [hunk 9]
```diff
@@ -34,7 +24,7 @@
       ></v-select>
     </v-row>
 
-    <div v-if="selected_axis==0">
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/collapse/collapse.vue:37]: If we're getting rid of the axis option entirely, then the `v-else` following this should also be removed (I see you already removed the `add_replace_results` traitlet that it uses.

Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/collapse.vue:37]: Good catch! Done.

**Comment by kecnry** [comment]
> While testing this, I noticed that collapsed data is in the wrong orientation.

This rotation does not seem to occur in main (when choosing axis=0).

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 10]
```diff
@@ -74,30 +62,28 @@ def _on_data_updated(self, msg):
 
     @observe('selected_data_item')
     def _on_data_item_selected(self, event):
-        self._selected_data = next((x for x in self.data_collection
-                                    if x.label == event['new']))
+        data_label = event['new']
+        if data_label not in self.data_collection.labels:
+            return
+        self._selected_data = self.data_collection[self.data_collection.labels.index(data_label)]
+        self._selected_cube = self._selected_data.get_object(cls=Spectrum1D, statistic=None)
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:69]: Note: I feel like there is no point to extract the cube multiple times across the plugin. Here, I change it to extracting it only once when data is selected.

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 11]
```diff
@@ -108,37 +94,26 @@ def vue_list_subsets(self, event):
...
+        spec_min = float(self.spectral_min) * u.Unit(self.spectral_unit)
+        spec_max = float(self.spectral_max) * u.Unit(self.spectral_unit)
 
-        collapsed_spec = getattr(spec, self.selected_func.lower())(
-            axis=self.selected_axis)
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore', message='No observer defined on WCS')
+            spec = spectral_slab(self._selected_cube, spec_min, spec_max)
+            # Spatial-spatial image only.
+            collapsed_spec = spec.collapse(self.selected_func.lower(), axis=-1).T  # Quantity
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/collapse.py:112]: Note: Transpose here fixes the orientation problem that @javerbukh found.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py [hunk 12]
```diff
@@ -1,56 +1,29 @@
...
-from jdaviz import Application
+from astropy import units as u
+from specutils import Spectrum1D
 
 from jdaviz.configs.default.plugins.collapse.collapse import Collapse
 
 
 @pytest.mark.filterwarnings('ignore')
-def test_linking_after_collapse(spectral_cube_wcs):
+def test_linking_after_collapse(cubeviz_helper, spectral_cube_wcs):
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/tests/test_collapse.py:10]: Note: Might as well use Cubeviz here to be more realistic.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py [hunk 13]
```diff
@@ -1,56 +1,29 @@
...
-    app = Application()
-    dc = app.data_collection
-    dc.append(Data(x=np.ones((3, 4, 5)), label='test', coords=spectral_cube_wcs))
+    coll = Collapse(app=cubeviz_helper.app)
 
-    coll = Collapse(app=app)
+    coll.selected_data_item = 'Unknown spectrum object[FLUX]'
 
-    coll.selected_data_item = 'test'
+    # This is automatically called in notebook but not in this test.
```
Comment (pllim) [inline jdaviz/configs/default/plugins/collapse/tests/test_collapse.py:18]: I don't know if this is frontend weirdness in unit test or a bug. Seems to work find in the plugin but please double check!

File: setup.cfg [hunk 14]
```diff
@@ -89,12 +88,12 @@ filterwarnings =
     ignore:distutils Version classes are deprecated:DeprecationWarning
     ignore:Passing unrecognized arguments to super:DeprecationWarning
     ignore:.*With traitlets 4\.1, metadata should be set using the \.tag\(\) method:DeprecationWarning
-    ignore::DeprecationWarning:spectral_cube
     ignore::DeprecationWarning:glue
     ignore::DeprecationWarning:bqplot
     ignore::DeprecationWarning:bqplot_image_gl
     ignore::DeprecationWarning:bqscales
     ignore::DeprecationWarning:traittypes
+    ignore::DeprecationWarning:voila
```
Comment (pllim) [inline setup.cfg:96]: I see a new warning locally, so I need this line so the tests would run on my machine.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py (hunk 13 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/collapse/tests/test_collapse.py:18]: Is the intention here to update the populated spectral bounds?  I think that's scheduled to be removed in the future anyways, but I am seeing it not being populated at start in this PR (but populated at start in main)...

File: jdaviz/configs/default/plugins/collapse/collapse.py [hunk 15]
```diff
@@ -74,30 +62,28 @@ def _on_data_updated(self, msg):
...
-        self.axes = list(range(len(self._selected_data.shape)))
+        self.selected_subset = "None"  # This calls self._on_subset_selected()
 
     @observe("selected_subset")
     def _on_subset_selected(self, event):
+        if self._selected_data is None:
+            return
+
         # If "None" selected, reset based on bounds of selected data
         if self.selected_subset == "None":
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/collapse/collapse.py:81]: Might be out-of-scope, but we've updated some of the plugins to use "Entire Spectrum" in place of "None".  We probably should just go through the remaining plugins and get them all to be consistent at some point soon.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py (hunk 13 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/tests/test_collapse.py:18]: Correct. Because it is my impression that this would automatically be picked up by `@observe`. So the way I refactored it, setting, `self.selected_data_item` actually calls `self.selected_subset` at the end, which I thought would kick off `self._on_subset_selected()` automatically to populate the bounds. Am I misunderstanding how `@observe` works?

File: jdaviz/configs/default/plugins/collapse/collapse.py (hunk 15 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/collapse.py:81]: I can change it here if there is consensus. If you do a code suggestion, you will get credit for it. If not, I'll change it at the next commit when I get a chance. Up to you.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py (hunk 13 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/collapse/tests/test_collapse.py:18]: Yes, except the default values set when creating the traitlets don't trigger those.  Your setup should work if you have the traitlet default to an empty string instead of `"None"` and let it update to `"None"` (and the bounds populate) when the data is loaded.

File: jdaviz/configs/default/plugins/collapse/collapse.py (hunk 15 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/collapse/collapse.py:81]: There are too many places to change it for a code suggestion, but I can add a commit if you'd prefer (but also don't really care about the credit if you want to just handle it in the same commit as https://github.com/spacetelescope/jdaviz/pull/1006#discussion_r803118536)

Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/collapse.py:81]: Oh, I mean just to change it for this plugin in this PR. But if you want to do it for the whole app, I agree that is a separate PR.

File: jdaviz/configs/default/plugins/collapse/tests/test_collapse.py (hunk 13 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/collapse/tests/test_collapse.py]: I wasn't aware of that subtlety. I pushed a commit. Please see if that's correct. Thanks!

**Comment by pllim** [comment]
Thanks for all the great reviews!
